### PR TITLE
add network driver based on the Virtio Over MMIO

### PR DESCRIPTION
**Pull request by stlankes**
Base on the work of @duanyu-yu, we developed a network driver based on the Virtio Over MMIO interface, which is part of the Virtual I/O Device Specification. Details are published at https://docs.oasis-open.org/virtio/virtio/v1.1/csprd01/virtio-v1.1-csprd01.html
    
This network driver doesn't depend on the PCI bus and can be used within micro VMs. The current version is developed and tested on Qemu's microvm virtual platform. Details are published at https://qemu.readthedocs.io/en/latest/system/i386/microvm.html
    
The current version can be tested with the example web server of RustyHermit's Git repository. Please use following command to build the web server:
    
> cargo build --no-default-features --features smoltcp -p httpd
    
Afterwards, the following command starts the web server withinthe microvm:
    
> qemu-system-x86_64 -M microvm,x-option-roms=off,pit=off,pic=off,rtc=on,auto-kernel-cmdline=off -global virtio-mmio.force-legacy=off -nodefaults -no-user-config -display none -smp 1 -m 1G -serial stdio -kernel loader/target/x86_64-unknown-hermit-loader/debug/rusty-loader -initrd target/x86_64-unknown-hermit/debug/httpd -enable-kvm -cpu host -device isa-debug-exit,iobase=0xf4,iosize=0x04 -netdev tap,id=tap10,ifname=tap10,script=no,downscript=no -device virtio-net-device,netdev=tap10

### CODE REVIEW

File: src/arch/x86_64/kernel/mmio.rs [hunk 1]
```diff
@@ -0,0 +1,148 @@
+// Copyright (c) 2021 Yu Duan, RWTH Aachen University
+//               2021, Stefan Lankes, RWTH Aachen University
+//
+// Licensed under the Apache License, Version 2.0, <LICENSE-APACHE or
+// http://apache.org/licenses/LICENSE-2.0> or the MIT license <LICENSE-MIT or
+// http://opensource.org/licenses/MIT>, at your option. This file may not be
+// copied, modified, or distributed except according to those terms.
+
```
Comment (mkroening) [inline src/arch/x86_64/kernel/mmio.rs:8]:
```suggestion
```

And add Yu Duan to `authors` in `Cargo.toml`.

File: src/drivers/net/virtio_mmio.rs [hunk 2]
```diff
@@ -0,0 +1,185 @@
+// Copyright (c) 2020 Frederik Schulz, RWTH Aachen University
+//
+// Licensed under the Apache License, Version 2.0, <LICENSE-APACHE or
+// http://apache.org/licenses/LICENSE-2.0> or the MIT license <LICENSE-MIT or
+// http://opensource.org/licenses/MIT>, at your option. This file may not be
+// copied, modified, or distributed except according to those terms.
+
```
Comment (mkroening) [inline src/drivers/net/virtio_mmio.rs:7]:
```suggestion
```

File: src/drivers/net/virtio_pci.rs [hunk 3]
```diff
@@ -0,0 +1,203 @@
+// Copyright (c) 2020 Frederik Schulz, RWTH Aachen University
+//
+// Licensed under the Apache License, Version 2.0, <LICENSE-APACHE or
+// http://apache.org/licenses/LICENSE-2.0> or the MIT license <LICENSE-MIT or
+// http://opensource.org/licenses/MIT>, at your option. This file may not be
+// copied, modified, or distributed except according to those terms.
+
```
Comment (mkroening) [inline src/drivers/net/virtio_pci.rs:7]:
```suggestion
```

File: src/drivers/virtio/transport/mmio.rs [hunk 4]
```diff
@@ -0,0 +1,614 @@
+// Copyright (c) 2021 Stefan Lankes, RWTH Aachen University
+//
+// Licensed under the Apache License, Version 2.0, <LICENSE-APACHE or
+// http://apache.org/licenses/LICENSE-2.0> or the MIT license <LICENSE-MIT or
+// http://opensource.org/licenses/MIT>, at your option. This file may not be
+// copied, modified, or distributed except according to those terms.
+
```
Comment (mkroening) [inline src/drivers/virtio/transport/mmio.rs:7]:
```suggestion
```

**Review by mkroening — CHANGES_REQUESTED** [review]
I found a few more copyright headers from Frederik Schulz and you, @stlankes.

I did not review the code in detail but in general this looks good to me.

r=me when copyright headers are gone.

**Comment by stlankes** [comment]
r=me

**Review by mkroening — APPROVED** [review]
bors r+

**Pull request merged**